### DISCUSSION TITLE
how can I use this CronExpression ecto type ?

### DISCUSSION THREAD

**Discussion by siddhant3030** [jshmrtn/crontab#84] [Q&A] [answered]
hi, I have one question. 

1:- 
 If I pass the cron_expression like this "* * * * *".  It's saving as a struct in my database. Something like this  ~e[*/21 * * * * *]. 
 But in order to consume from the frontend, I need to convert this struct to some other type. 
 
 So I tried this

1. CronExpression.to_condition_list(availability.cron_expression) and this will give me something like this

```
[
  minute: [{:/, :*, 21}],
  hour: [:*],
  day: [:*],
  month: [:*],
  weekday: [:*],
  year: [:*]
]
```

but I'm not able to pass this in a JSON. In order to consume this, what data should I send? What should I do with struct? 

Right now I'm getting this error when I'm trying to send after using the `to_condition_list` function. 

`** (Protocol.UndefinedError) protocol Jason.Encoder not implemented for {:minute, [{:/, :*, 21}]} of type Tuple, Jason.Encoder protocol must always be explicitly implemented.`

**Comment by maennchen** [comment, marked as answer]
@siddhant3030 You'll have to send it to the frontend as a string or implement a Jason.Encoder for the struct.

https://hexdocs.pm/jason/Jason.Encoder.html

**Reply by siddhant3030** [reply]
Thanks. But I tried that 

Even if I do that suppose I have a map something like this 

```
%{
  "day" => [:*],
  "extended" => false,
  "hour" => [:*],
  "minute" => [{:/, :*, 21}],
  "month" => [:*],
  "reboot" => false,
  "second" => [:*],
  "weekday" => [:*],
  "year" => [:*]
}
```
Even if I want to try to encode this Jason.encode!(b)

(Protocol.UndefinedError) protocol Jason.Encoder not implemented for {:/, :*, 21} of type Tuple

Now in every case there will be a different expression. how do I deal with all the possibilities?

**Reply by maennchen** [reply]
That is expected. Jason is not able to encode a tuple since it has no equivalent in JSON.

Therefore either:

1. Use string. Since this is how a corn expression is written, that's probably the way to go.
2. Transform it to JSON manually. This means that you will have to map all possible values to JSON and write a function that does this.

**Reply by siddhant3030** [reply]
Does that mean I don't need to use this ecto type? because that automatically stores the struct in the database? 
I mean
how do I store the string? bec I need to parse it before storing it and parsing it
 will give me a struct

**Reply by maennchen** [reply]
@siddhant3030 What does this have to do with JSON?

Ecto is for the database. There's no need to do anything manually with JSON when using Ecto.

**Reply by maennchen** [reply]
Which problem are you trying to solve?

* Frontend / Backend communication
* Saving data in the db

**Reply by siddhant3030** [reply]
I'm trying to solve frontend/backend communication. Since I'm using Ecto.Type and its storing as a struct. Crontab.CronExpression.Ecto.Type. This one. 

Since I have to pass this in a JSON. I need to encode this somehow.

**Reply by maennchen** [reply]
@siddhant3030 then why don't you just convert the struck to a string before you pass it to the frontend?

**Reply by siddhant3030** [reply]
That's what I was trying to do and where I got those errors. I can't change struct to a string. I will get the same issue with the tuple. 

I tried something like this

```
  # cron = Map.from_struct(cron_expression)
    # a = Map.new cron, fn {k, v} -> {to_string(k), v} end
    # IO.inspect(a)
    # cron_expression = Jason.encode!(a)
    # IO.inspect(cron_expression)
```

**Reply by maennchen** [reply]
Again: Is there a reason why you do not just serialize a string?

Just put `Crontab.CronExpression.Composer.compose(cron_expression)` into the JSON...

**Reply by siddhant3030** [reply]
Thankyou for help :)
